Flink:  Add a rewrite datafile action for flink

zhangjun0x01 opened the pull request:
this pr is for #1610 

Add a rewrite datafile action for Flink, the PR contains the content of #1624. PR #1624 extracts some common functions to iceberg core. I will do the rebase after PR #1624 is merged.

simonsssu commented on core/src/main/java/org/apache/iceberg/actions/RewriteDataFilesActionBase.java:
```diff
@@ -0,0 +1,271 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class RewriteDataFilesActionBase<ThisT, R> extends BaseSnapshotUpdateAction<ThisT, R> {
```
Hi ZhangJun, Thanks for your work. Seems the class only used in Flink and the logic here already exists in Spark part. I think it's better to split this pr into 2 parts, one is refactor current rewrite code and the other is 'add a rewrite action for flink'.

zhangjun0x01 replied:
hi,simonsssu.thanks for your suggestion. I have opened a new PR #1624  and extracted some common logic. I will do rebase in PR #1624 is merged

simonsssu commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,193 @@
+      this.taskWriterFactory.initialize(subTaskId, attemptId);
+
+    }
+
```
Useless blank line.

simonsssu commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,193 @@
+    public List<DataFile> map(CombinedScanTask task) throws Exception {
+      // Initialize the task writer.
+      this.writer = taskWriterFactory.create();
+      RowDataIterator iterator =
```
Can we use a try-with-resource here?  I think it will simply the code to handle exception.

zhangjun0x01 replied:
I refer to the spark code, but I didn't find a specific reason for doing so, so I copied it to here . I guess it may have encountered some special exceptions.

openinx commented:
Thanks @zhangjun0x01 and @simonsssu  for the work ,  I will take a look today.

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,193 @@
+        }
+        iterator.close();
+        writer.close();
+        return Lists.newArrayList(writer.complete()).stream()
```
Could we just return  the `Lists.newArrayList(writer.complete())` ?

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,193 @@
+          writer.write(rowData);
+        }
+        iterator.close();
+        writer.close();
```
The following `writer.complete()` will close firstly and then return the completed data file list, pls see the javadoc in `TaskWriter`: 

```java
  /**
   * Close the writer and get the completed data files.
   *
   * @return the completed data files of this task writer.
   */
  DataFile[] complete() throws IOException;
```
So here we don't have to call `writer.close()` now.

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,193 @@
+
+      // Initialize the task writer factory.
+      this.taskWriterFactory.initialize(subTaskId, attemptId);
+
```
nit:  pls don't left a new empty line here.

zhangjun0x01 replied:
I use the iceberg checkstyle to format the code,Is it necessary to add this empty line  limit to checkstyle?

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,193 @@
+  private static final Logger LOG = LoggerFactory.getLogger(RowDataRewriter.class);
+
+  private final Schema schema;
+  private final PartitionSpec spec;
```
Those variables `spec`, `properties`, `locations` are never used in this classes , do we need them here ?

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,193 @@
+  private final FileIO io;
+  private final boolean caseSensitive;
+  private final EncryptionManager encryptionManager;
+  private final Table table;
```
The same thing for `table`.

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,193 @@
+    this.caseSensitive = caseSensitive;
+    this.nameMapping = table.properties().get(DEFAULT_NAME_MAPPING);
+
+    String formatString = table.properties().getOrDefault(
```
Maybe it's time to move all the properties parser to a `TablePropsUtil` now.

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,193 @@
+        table.locationProvider(),
+        table.io(),
+        table.encryption(),
+        Long.MAX_VALUE,
```
The `targetFileSizeBytes` will always be unlimited ?  Pls set a configuration keys for it,  [here](https://github.com/apache/iceberg/pull/1083/files#r438890925) is the discussion we had for spark compaction.

zhangjun0x01 replied:
I have read this discussion, but my understanding is that the limit of targetFileSizeBytes is in the reader, not the writer. 
the writer of the spark is [here](https://github.com/apache/iceberg/blob/master/spark/src/main/java/org/apache/iceberg/spark/source/RowDataRewriter.java#L105)

```
    TaskWriter<InternalRow> writer;
    if (spec.fields().isEmpty()) {
      writer = new UnpartitionedWriter<>(spec, format, appenderFactory, fileFactory, io.value(), Long.MAX_VALUE);
    } else {
      writer = new SparkPartitionedWriter(spec, format, appenderFactory, fileFactory, io.value(), Long.MAX_VALUE,
          schema, structType);
    }
```

I set the targetFileSizeBytes in [RewriteDataFilesActionBase#targetSizeInBytes method ](https://github.com/zhangjun0x01/iceberg/blob/RewriteDataFilesAction/core/src/main/java/org/apache/iceberg/actions/RewriteDataFilesActionBase.java#L200)

openinx replied:
The conclusion should be [here](https://github.com/apache/iceberg/pull/1083/files#r439180961).  OK,  we only have to control the read split size and ignore the write target file size.

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,193 @@
+          LOG.error("Aborting task", originalThrowable);
+
+          LOG.error("Aborting commit for  (subTaskId {}, attemptId {})",
+              subTaskId, attemptId);
```
nit: Do we need to switch to a new line here ?

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,193 @@
+
+          LOG.error("Aborting commit for  (subTaskId {}, attemptId {})",
+              subTaskId, attemptId);
+          if (iterator != null) {
```
nit:  iterator is always not-null ?

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,193 @@
+
+import static org.apache.iceberg.TableProperties.DEFAULT_NAME_MAPPING;
+
+public class RowDataRewriter implements Serializable {
```
Does this class need to be serializable ?  I checked the code, seems it won't need to be encoded or decoded for network transferring ?

openinx commented on flink/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,91 @@
+    try {
+      addedDataFiles = rowDataRewriter.rewriteDataForTasks(dataSet);
+    } catch (Exception e) {
+      LOG.error("rewrite data files error", e);
```
I think here we'd better to throw the exception to upper layer ?  How about throwing the `UncheckedIOException(e)` ?

zhangjun0x01 replied:
This exception comes from the DataSet#collect method. I looked at this method again. It throws a RuntimeException. I think we also throw a RuntimeException here. What do you think?

openinx replied:
Both `RuntimeException`  and `UncheckedIOException` works fine for me. Actually,  `UncheckedIOException` is also a `RuntimeException`.

zhangjun0x01 commented:
Thank you very much for the review @openinx @simonsssu , I have solved all comments

zhangjun0x01 commented on flink/src/main/java/org/apache/iceberg/flink/TablePropertiesUtil.java:
```diff
@@ -0,0 +1,38 @@
+/*
```
yes,we should use PropertyUtil, I had delete the TablePropertiesUtil.java ,I didn't find this PropertyUtil at first

openinx commented:
@zhangjun0x01 Could you rebase this patch to resolve the conflicts ?  I will take another look for this, Thanks.

zhangjun0x01 commented:
@openinx I updated the patch.

As discussed in #1624, the variable fileIO in BaseRewriteDataFilesAction lacks a getter method, so I used BaseRewriteDataFilesAction#fileIO() as its getter method and added a setter method setFileIO().

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,153 @@
+
+  public static class RewriteMap extends RichMapFunction<CombinedScanTask, List<DataFile>> {
+
+    private transient TaskWriter<RowData> writer;
```
nit:  could just use local variable ?  don't have to be a transient field.

openinx commented on flink/src/test/java/org/apache/iceberg/actions/TestRewriteDataFilesAction.java:
```diff
@@ -0,0 +1,211 @@
+    Assert.assertEquals("Should have 4 data files after rewrite", 4, dataFiles1.size());
+
+    // Assert the table records as expected.
+    SimpleDataUtil.assertTableRecords(icebergTablePartitioned, Lists.newArrayList(
```
I think we may need to refactor the `SimpleDataUtil.assertTableRecords`, because it will use `Set` to check the expected records, that means it will de-duplicate the same records automatically, and won't identify the case that have two `<1, 'hello'>` records.

openinx commented on flink/src/test/java/org/apache/iceberg/flink/FlinkTestBase.java:
```diff
@@ -72,7 +72,7 @@ protected TableEnvironment getTableEnv() {
   }
 
-  List<Object[]> sql(String query, Object... args) {
+  public List<Object[]> sql(String query, Object... args) {
```
How about use `protected` ?

openinx commented on flink/src/test/java/org/apache/iceberg/actions/TestRewriteDataFilesAction.java:
```diff
@@ -0,0 +1,211 @@
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestRewriteDataFilesAction extends FlinkCatalogTestBase {
```
I think we should also provide an unit test similar to the spark `testRewriteLargeTableHasResiduals`, that test the option `ignoreResiduals`.

zhangjun0x01 replied:
done

zhangjun0x01 commented on core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:
```diff
@@ -268,7 +272,7 @@ private void replaceDataFiles(Iterable<DataFile> deletedDataFiles, Iterable<Data
   }
 
-  protected abstract FileIO fileIO();
+  protected abstract FileIO setFileIO();
```
In the construction method of BaseRewriteDataFilesAction, we need to initialize fileIO and provide an abstract setter method. Other places, such as the rewriteDataForTasks method of spark and flink, need to get fileIO by call the getter method of fileIO, but there is no getter method currently. If we also call the fileIO() method to get fileIO, then the fileIO() method will be called multiple times, and SparkUtil.serializableFileIO() will be called multiple times.

We provide a setFileIO() method to ensure that the initialization of fileIO is only executed once. If we want to get the value of the fileIO field, call its getter method fileIO().

openinx replied:
I understand your points. I said it's confusing because its method name is `setFileIO` while would just return the `FileIO` without accepting any arguments,   it looks like a `getFileIO`. 

How about loading the fileIO lazily as the previous abstraction PR comment ? 

```java
private FileIO fileIO;

private FileIO fileIO(){
    if(fileIO==null){
       this.fileIO = SparkUtil.serializableFileIO(table());
    } 
    return this.fileIO;
}

```

openinx commented on core/src/main/java/org/apache/iceberg/util/PropertyUtil.java:
```diff
@@ -52,4 +52,13 @@ public static long propertyAsLong(Map<String, String> properties,
   }
+
+  public static String propertyAsString(Map<String, String> properties,
+                                    String property, String defaultValue) {
```
I mean the `String property` should be aligned with the `Map<String, String> properties`. 

```java
  public static String propertyAsString(Map<String, String> properties,
                                        String property, String defaultValue) {
    String value = properties.get(property);
    if (value != null) {
      return properties.get(property);
    }
    return defaultValue;
  }
```

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,153 @@
+
+import static org.apache.iceberg.TableProperties.DEFAULT_NAME_MAPPING;
+
+public class RowDataRewriter {
```
OK, we've put the `RowDataRewriter` in the `org.apache.iceberg.flink.source` package, while the `RewriteDataFilesAction` is in `org.apache.iceberg.actions` package.  So we need to make it public, sounds reasonable.

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,153 @@
+        iterator.close();
+        return Lists.newArrayList(writer.complete());
+      } catch (Throwable originalThrowable) {
+        try {
```
nit: I think it's good to use the `try-with-resource` here as @simonsssu  suggested. 

```java
      try (RowDataIterator iterator = new RowDataIterator(task, io, encryptionManager, schema, schema, nameMapping,
          caseSensitive)) {
        while (iterator.hasNext()) {
          RowData rowData = iterator.next();
          writer.write(rowData);
        }
        return Lists.newArrayList(writer.complete());
      } catch (Throwable originalThrowable) {
        try {
          LOG.error("Aborting commit for  (subTaskId {}, attemptId {})", subTaskId, attemptId);
          writer.abort();
          LOG.error("Aborted commit for  (subTaskId {}, attemptId {})", subTaskId, attemptId);
        } catch (Throwable inner) {
          if (originalThrowable != inner) {
            originalThrowable.addSuppressed(inner);
            LOG.warn("Suppressing exception in catch: {}", inner.getMessage(), inner);
          }
        }
        if (originalThrowable instanceof Exception) {
          throw originalThrowable;
        } else {
          throw new RuntimeException(originalThrowable);
        }
      }
    }
```

openinx reviewed (commented):
That patch almost looks good to me, left several minor comments.  Thanks @zhangjun0x01 for the work.

openinx commented on flink/src/test/java/org/apache/iceberg/flink/actions/TestRewriteDataFilesAction.java:
```diff
@@ -0,0 +1,283 @@
+@RunWith(Parameterized.class)
+public class TestRewriteDataFilesAction extends FlinkCatalogTestBase {
+
+  private static final String TABLE_NAME_UNPARTIITONED = "test_table_unpartitioned";
```
nit: TABLE_NAME_UNPARTIITONED -> `TABLE_NAME_UNPARTITIONED`. 

`TABLE_NAME_PARTIITONED` has the similar issue.

zhangjun0x01 replied:
I updated it

openinx reviewed (approved):
I think it's ready to go,  @rdblue would you like to have a double check ?

JingsongLi reviewed (commented):
Thanks @zhangjun0x01 for your contribution, why do you want to use `DataSet` API? `DataSet` API is being deprecated.
Looks like just need a `parallelize` and `map` and `collect`. I think `DataStream`(We can call it BoundedStream) can finish this.
What do you think?

JingsongLi commented on flink/src/main/java/org/apache/iceberg/flink/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,63 @@
+
+  @Override
+  protected List<DataFile> rewriteDataForTasks(List<CombinedScanTask> combinedScanTasks) {
+    DataSet<CombinedScanTask> dataSet = env.fromCollection(combinedScanTasks).setParallelism(combinedScanTasks.size());
```
I think `setParallelism` may not works.
What we want is a `SparkContext.parallelize`, but this `setParallelism` just set parallelism, records still in a single node, downstream operators only have computation on one node.
Can you check this?

zhangjun0x01 replied:
> Thanks @zhangjun0x01 for your contribution, why do you want to use `DataSet` API? `DataSet` API is being deprecated.
> Looks like just need a `parallelize` and `map` and `collect`. I think `DataStream`(We can call it BoundedStream) can finish this.
> What do you think?

Hi,@JingsongLi：
I look up the api of the DataStream, it seems  no `collect` method

JingsongLi replied:
You can use `DataStreamUtils.collect`

zhangjun0x01 replied:
I updated the pr, use DataStream instead of DataSet , and set the parallelism to StreamExecutionEnvironment.

but setParallelism for  `env.fromCollection(combinedScanTasks)`  does not seem to have an impact on downstream operator,this is my test job graph for use dataset.   https://blog.csdn.net/zhangjun5965/article/details/109469798

JingsongLi replied:
Thanks for the update and check, I think you are right, the graph is good.

JingsongLi replied:
I think about parallelism again.
Can we not set parallelism? Flink is a push based model, which is not suitable for taking on too much parallelism.
So can we just let it as default parallelims?

zhangjun0x01 replied:
If we do not set the parallelism, we can use the default parallelism of flink. If the user wants to modify the parallelism to improve the execution speed, it can be set by the `-p` parameter. However, the user may not know how much parallelism to set，because users may not know how many data files to Rewrite each time. If the setting is too large, it may cause a waste of resources. 

So I think we can  set a max parallelism. If the size of CombinedScanTask list  is less than max parallelism, we use the CombinedScanTask list size as the job parallelism, if the size  is greater than max parallelism，we use the max parallelism which we set.
What do you think?

JingsongLi replied:
Let me explain why we can't set parallelism directly:
Flink is in pipeline mode by default, which means that all tasks must be started.
If we want to compact a large table, it may require thousands of parallelism, which leads to the need for a very large cluster.

This is similar to the Hive source parallelism inference: https://ci.apache.org/projects/flink/flink-docs-master/dev/table/hive/hive_read_write.html#source-parallelism-inference
The value should be used to infer parallelism, but we need a max parallelism.

So I agree with you very much. Let's add a max parallelism.

JingsongLi replied:
BTW, you should set parallelism after `map` instead of modifying env.

zhangjun0x01 replied:
I updated the pr, although we set the parallelism for StreamExecutionEnvironment, I found that the DataStreamUtils.collect operator only uses 1 parallelism, so set parallelism to map operator or StreamExecutionEnvironment seems to have the same effect, but set parallelism to map is better to understand .

I submit an issue (#1660) to  set max parallelism a few days ago, I will open a new pr  for #1660 when this PR is merged

JingsongLi replied:
> although we set the parallelism for StreamExecutionEnvironment, I found that the DataStreamUtils.collect operator only uses 1 parallelism, so set parallelism to map operator or StreamExecutionEnvironment seems to have the same effect, but set parallelism to map is better to understand.

It is better not to bring extra impact to env, because this env may be reused by users.

> I submit an issue (#1660) to set max parallelism a few days ago, I will open a new pr for #1660 when this PR is merged

Can you modify in this PR? Or just not set parallelism in this PR?
I think it is not good to bring user cluster risk to master code.

zhangjun0x01 replied:
I  update the code , add the max parallelism in this PR.

JingsongLi commented:
BTW, @zhangjun0x01 It is better to modify PR title to `Flink: Add a rewrite datafile action`

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataRewriter.java:
```diff
@@ -0,0 +1,155 @@
+        table.properties());
+  }
+
+  public List<DataFile> rewriteDataForTasks(DataStream<CombinedScanTask> dataSet, int parallelism) {
```
If change to use datastream to submit this job,  then the argument `dataSet` should be renamed to `dataStream` ?

zhangjun0x01 replied:
I updated it.

openinx commented on core/src/main/java/org/apache/iceberg/actions/BaseRewriteDataFilesAction.java:
```diff
@@ -268,6 +268,23 @@ private void replaceDataFiles(Iterable<DataFile> deletedDataFiles, Iterable<Data
     }
   }
 
+  protected int getParallelism(int parallelism, List<CombinedScanTask> combinedScanTasks) {
```
Do we need to pass the whole `List<CombinedScanTask>` here ?  I think we only need the `numOfScanTasks` ?

openinx replied:
btw,  it seems to be a method which was designed for flink, Better to move it to flink implementation class ?

JingsongLi replied:
+1 I think we can move options to Flink too.

zhangjun0x01 replied:
should we set the max parallelism for spark  RewriteDataFilesAction?

JingsongLi replied:
Spark can handle large parallelism in default...

zhangjun0x01 replied:
I moved set max parallelism to flink RewriteDataFilesAction

openinx commented:
@JingsongLi   Any other concern for this patch ?

JingsongLi commented:
Hi @zhangjun0x01 , I think we can add a method `setMaxParallelism` to Flink `RewriteDataFilesAction`. It don't need be a table property.
And I think the default value of maxParallelism can be `env.getParallelism`

JingsongLi reviewed (approved):
Thanks for the update, looks good to me.
